File: db-migration-scripts/create_migration.js
/**
 * Requires NodeJS >= 18
 * --------------------------------------------------
 * Run with
 * $ node create_migration.js
 * --------------------------------------------------
 * 
 * This script creates files based in utc timestamp granting liabilty when creating database migrations.
 * By default, the file will be empty and placed inside a directory called migration.
 * Those must be in the same directory level as the script.
 * 
 * If you want to create a rollback for this migration, use the create_migration_and_rollback.js script instead.
 * If you want to change the output directory, change MIGRATION_OUTPUT_FOLDER.
 * 
 * Just enter a name you want when asked, and it will be created with the following pattern:
 * OUTPUT_FILE_PREFIX + CURRENT_UTC_TIMESTAMP + "__" + MIGRATION_FILE_NAME + OUTPUT_FILE_SUFFIX
 */
const readline = require('node:readline/promises');
const fs = require('fs/promises');
const fsSync = require('fs');
const path = require('path');
const { stdin, stdout } = require('node:process');

// SETTINGS START
const MIGRATION_OUTPUT_FOLDER = './migration';
const OUTPUT_FILE_PREFIX = 'V';
const CURRENT_UTC_TIMESTAMP = Math.trunc(new Date().getTime() / 1000);
const OUTPUT_FILE_SUFFIX = '.sql';
// SETTINGS END

const run = async () => {
  const rl = readline.createInterface({ input: stdin, output: stdout });
  console.log('--------------------------------------------------');
  console.log('CREATE MIGRATION FILE');
  console.log('--------------------------------------------------');
  console.log('Type a name for your file (blank spaces are not supported):');
  console.log('---- Use snake_case_in_your_name to separate words.');
  console.log('---- SIGINT (CTRL + C) to cancel.');

  const migrationPath = path.join(__dirname, MIGRATION_OUTPUT_FOLDER);
  const branchName = await rl.question('');

  rl.close();

  const migrationFileName = `${OUTPUT_FILE_PREFIX}${CURRENT_UTC_TIMESTAMP}__${branchName}${OUTPUT_FILE_SUFFIX}`;

  console.log(`Creating migration output folder ${MIGRATION_OUTPUT_FOLDER} (if not exists)`)
  if (!fsSync.existsSync(migrationPath)) {
    fsSync.mkdirSync(migrationPath, { recursive: true });
  }
  
  console.log(`Creating migration file ${migrationFileName} in ${MIGRATION_OUTPUT_FOLDER}`)
  const migrationFileCreationPath = path.join(migrationPath, migrationFileName);
  await fs.writeFile(migrationFileCreationPath, '');
  console.log('--------------------------------------------------');
  console.log('Finished creating migration file.');
  console.log('--------------------------------------------------');
}

run().catch((err) => {
  console.error('Error creating migration');
  console.error(err);
});

File: scripts/db-migration/create_migration.py
#
# Requires Python 3+
# --------------------------------------------------
# Run with
# $ python3/python/py create_migration.py
# --------------------------------------------------
# 
# This script creates files based in utc timestamp granting liability when creating database migrations.
# By default, the file will be empty and placed inside a directory called migration.
# Those must be in the same directory level as the script.
# 
# If you want to create a rollback for this migration, use the create_migration_and_rollback.py script instead.
# If you want to change the output directory, change MIGRATION_OUTPUT_FOLDER.
# 
# Just enter a name you want when asked, and it will be created with the following pattern:
# OUTPUT_FILE_PREFIX + CURRENT_UTC_TIMESTAMP + "__" + MIGRATION_FILE_NAME + OUTPUT_FILE_SUFFIX
# 

import time, os

# SETTINGS START
MIGRATION_OUTPUT_FOLDER = './migration'
OUTPUT_FILE_PREFIX = 'V'
CURRENT_UTC_TIMESTAMP = int(time.time())
OUTPUT_FILE_SUFFIX = '.sql'
# SETTINGS END

def main():
    print("--------------------------------------------------")
    print("CREATE MIGRATION FILE")
    print("--------------------------------------------------")
    print("Type a name for your file (blank spaces are not supported):")
    print("---- Use snake_case_in_your_name to separate words.")
    print("---- SIGINT (CTRL + C) to cancel.")
    branch_name = input("")

    migration_file_name = f'{OUTPUT_FILE_PREFIX}{CURRENT_UTC_TIMESTAMP}__{branch_name}{OUTPUT_FILE_SUFFIX}'
    migration_file_path = f'{MIGRATION_OUTPUT_FOLDER}/{migration_file_name}'
    
    print(f'Creating migration output folder {MIGRATION_OUTPUT_FOLDER} (if not exists)')
    os.makedirs(os.path.dirname(migration_file_path), exist_ok=True)

    print(f'Creating migration file {migration_file_name} in {MIGRATION_OUTPUT_FOLDER}')
    with open(migration_file_path, 'w') as file:
        file.write('');

    print("--------------------------------------------------")
    print("Finished creating migration file.")
    print("--------------------------------------------------")

main()
